api/v1alpha1/amaltheasession_types.go

```diff
@@ -87,8 +89,19 @@ type AmaltheaSessionSpec struct {
 	// +optional
 	Tolerations []v1.Toleration `json:"tolerations,omitempty"`
 
-	//Control over the sidecars and init containers Amalthea adds to each session
-	Sidecars Sidecars `json:"sidecars,omitempty"`
+	// +kubebuilder:default:="always"
+	// Indicates how Amalthea should reconcile the child resources for a session. This can be problematic because
+	// newer versions of Amalthea may include new versions of the sidecars or other changes not reflected
+	// in the AmaltheaSession CRD, so simply updating Amalthea could cause existing sessions to restart
+	// because the sidecars will have a newer image or for other reasons because the code changed.
+	// Hibernating the session and deleting it will always work as expected regardless of the strategy.
+	// The status of the session and all hibernation or auto-cleanup functionality will always work as expected.
+	// A few values are possible:
+	// - never: Amalthea will never update any of the child resources and will ignore any changes to the CR
+	// - always: This is the expected method of operation for an operator, changes to the spec are always reconciled
```

Comment:
Is there a way to notify the user that the session needs to be restarted ? Otherwise we will have the issue of work getting interrupted at the least good moment.

Reply:
@sgaist we will most likely not be running Amalthea with the `always` strategy in renku. I am not sure who you are referring to by "user" because the only "user" we can reliably notify from Amalthea is the admin/service that created the session in k8s. And we can do this via the status field in the CR. But in the case of `always`, notifying the user that the session needs to be restarted makes no sense. The whole point of this strategy is that things will just happen immediately without action from the user or admin.

We can add a notification that the session will be updated with the `whenHibernatedOrFailed` strategy. So that users and admins know that the session is out of date and will be updated when the session is hibernated.

We can do a similar thing for the `never` strategy. 

With always what we can try to do is add something that indicates that the session has been updated. 

I will do this in a follow up PR.

Reply:
See #749